os/arch/arm/src/armv7-a/arm_assert.c

```diff
@@ -497,61 +479,39 @@ static inline void print_assert_detail(const uint8_t *filename, int lineno, stru
-	lldbg_noarg("=====================================================================\n");
-	if (mm_check_heap_corruption(g_kmmheap) == OK) {
-		lldbg_noarg("No kernel heap corruption detected\n");
-	} else {
-		lldbg_noarg("##########################################################################################################################################\n");
-		lldbg_noarg("\n\n");
-		/* treat kernel fault */
-		arm_assert();
-	}
-	lldbg_noarg("=====================================================================\n");
-	lldbg_noarg("\n");
+	up_dumpstate(fault_tcb, asserted_location);
 	
 	/* Dump the state of all tasks (if available) */
 	task_show_alivetask_list();
-	lldbg_noarg("\n");
 
-	/* Dump the asserted TCB */
-	lldbg_noarg("=====================================================================\n");
-	lldbg_noarg("Asserted TCB Info\n");
-	lldbg_noarg("=====================================================================\n");
-	task_show_tcbinfo(fault_tcb);
-	lldbg_noarg("=====================================================================\n");
-	lldbg_noarg("\n");
 
-#ifdef CONFIG_APP_BINARY_SEPARATION
-	lldbg_noarg("=====================================================================\n");
-	lldbg_noarg("Application Info\n");
-	lldbg_noarg("=====================================================================\n");
-	elf_show_all_bin_section_addr();
-	lldbg_noarg("=====================================================================\n");
-	lldbg_noarg("\n");
-#endif
-
-	up_dumpstate(fault_tcb, asserted_location);
 
 #ifdef CONFIG_APP_BINARY_SEPARATION
+	elf_show_all_bin_section_addr();
 	if (IS_FAULT_IN_USER_THREAD(fault_tcb)) {
-		lldbg_noarg("=====================================================================\n");
-		lldbg_noarg("Checking current app heap for corruption...\n");
-		lldbg_noarg("=====================================================================\n");
+		lldbg_noarg("===========================================================App_Heap_Corruption==========================================\n");
+		lldbg("Checking current app heap for corruption...\n");
 		if (mm_check_heap_corruption((struct mm_heap_s *)(fault_tcb->uheap)) == OK) {
-			lldbg_noarg("No app heap corruption detected\n");
+			lldbg("No app heap corruption detected\n");
 		}
-		lldbg_noarg("=====================================================================\n");
 	}
 #endif
 
-#if defined(CONFIG_BOARD_CRASHDUMP)
-	lldbg_noarg("Perform Crashdump\n");
-	board_crashdump(up_getsp(), fault_tcb, (uint8_t *)filename, lineno);
-	lldbg_noarg("\n");
-#endif
+	lldbg_noarg("===========================================================Kernel_Heap_Corruption==========================================\n");
+	lldbg("Checking kernel heap for corruption...\n");
+	if (mm_check_heap_corruption(g_kmmheap) == OK) {
+		lldbg("No kernel heap corruption detected\n");
+	} else {
+		/* treat kernel fault */
+		arm_assert();
+	}
```

Comment:
1. Is this a kind of dump state?
2. This code is called always.
    1. in dump state function with secured
    2. in up_assert without secured

Because of two reasons above, how about calling it in up_assert once regardless of secured?

Reply:
How about seperate heap check code to new function in up_assert()?

```
void check_heap_corrupt() {
	if (!IS_SECURE_STATE()) {
		//user heap check and logs
		//logs of kernel heap check and trap parsing logs
	} 
	// kernel heap check
}

void up_assert() {
	...
	check_heap_corrupt();
}

```

Reply:
Thank you for the suggestion. Looks good to me.
Then, we need to shift Task TCB Info and closing log line to up_assert function.
Can it be implement this way?
```
void check_heap_corrupt() {
	if (!IS_SECURE_STATE()) {
		//user heap check and logs
		//logs of kernel heap check and trap parsing logs
	} 
	if (mm_check_heap_corruption(g_kmmheap) != OK) {
		/* treat kernel fault */
		arm_assert();
	} else {
		if (!IS_SECURE_STATE()) {
			lldbg("No kernel heap corruption detected\n");
		}		
	}
}

void up_assert() {
	...
	check_heap_corrupt();
	if (!IS_SECURE_STATE()) {
		* Dump the asserted TCB */
		task_show_tcbinfo(fault_tcb);

		/* Closing log line */
		lldbg_noarg("##########################################################################################################################################\n");
	}
}
```

Reply:
I think would be okay to output the heap check log (user and kernel heap) at the end of the entire log.

and you can move `lldbg("No kernel heap corruption detected\n");` to in mm_check_heap_corruption() function.

Reply:
It would be okay to put closing log line before heap check log?

Reply:
The closing log line must locate last.

but we don't need to check os secure state when put closing line